src/main/java/spoon/support/util/ModelList.java

```diff
@@ -0,0 +1,245 @@
+ * and/or redistribute the software under the terms of the CeCILL-C license as
+ * circulated by CEA, CNRS and INRIA at http://www.cecill.info.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the CeCILL-C License for more details.
+ *
+ * The fact that you are presently reading this means that you have had
+ * knowledge of the CeCILL-C license and that you accept its terms.
+ */
+package spoon.support.util;
+
+
+import java.io.Serializable;
+import java.util.AbstractList;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.ConcurrentModificationException;
+import java.util.List;
+
+import spoon.SpoonException;
+import spoon.experimental.modelobs.FineModelChangeListener;
+import spoon.reflect.declaration.CtElement;
+import spoon.reflect.path.CtRole;
+import spoon.support.reflect.declaration.CtElementImpl;
+
+/**
+ * The implementation of the {@link List}, which is used by Spoon model objects.
+ * It assures:
+ * 1) each inserted {@link CtElement} gets assigned correct parent
+ * 2) each change is reported in {@link FineModelChangeListener}
+ */
+public abstract class ModelList<T extends CtElement> extends AbstractList<T> implements Serializable {
+	private static final long serialVersionUID = 1L;
+
+	private List<T> list = CtElementImpl.emptyList();
+
+	protected ModelList() {
+	}
+
+	protected abstract CtElement getOwner();
+	protected abstract CtRole getRole();
+	protected abstract int getDefaultCapacity();
+	protected void onSizeChanged(int newSize) {
+	}
+
+	@Override
+	public T get(int index) {
+		return list.get(index);
+	}
+
+	/**
+	 * sets the new content of this List
+	 * @param elements new content of this list
+	 */
+	public void set(Collection<T> elements) {
+		//TODO the best would be to detect added/removed statements and to fire modifications only for them
+		this.clear();
+		if (elements != null && elements.isEmpty() == false) {
+			this.addAll(elements);
+		}
+	}
+
+	@Override
+	public int size() {
+		return list.size();
+	}
+
+	@Override
+	public T set(int index, T element) {
+		T oldElement = list.get(index);
+		if (oldElement == element) {
+			//no change
+			return oldElement;
+		}
+		CtElement owner = getOwner();
+		ensureModifiableList();
+		getModelChangeListener().onListDelete(owner, getRole(), list, index, oldElement);
+		linkToParent(owner, element);
+		getModelChangeListener().onListAdd(owner, getRole(), list, index, element);
+		list.set(index, element);
+		updateModCount();
+		return oldElement;
+	}
+
+	static void linkToParent(CtElement owner, CtElement element) {
+		if (owner.getFactory().getEnvironment().checksAreSkipped() == false && element.isParentInitialized() && element.getParent() != owner) {
+			//the `e` already has an different parent. Check if it is still linked to that parent
+			if (element.getRoleInParent() != null) {
```

Comment:
This exception was essential for finding errors. But should we keep it now? It will break a lot of client code.

Reply:
it is already configurable here `owner.getFactory().getEnvironment().checksAreSkipped()` so one can switch it off

> break a lot of client code.

It breaks only client code which is already broken, because that code creates invalid AST, which can cause unexpected behavior of spoon and tricky issues.